cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -126,42 +127,86 @@ static void show_buffers(const char *line, int offset,
 	fprintf(stderr, ">%s<\n", line);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i) {
+	for (size_t i = 0; i < strlen(line); ++i) {
 		int c = entry->line_buf[i];
 		fputc(c == 0 ? '@' : c == 1 ? '#' : c, stderr);
 	}
 	fputc('<', stderr);
 	fputc('\n', stderr);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i)
+	for (size_t i = 0; i < strlen(line); ++i)
 		fputc('=', stderr);
 	fputc('<', stderr);
 	fputc('\n', stderr);
 #endif				// MOUNTINFO_DEBUG
 }
 
+static bool is_octal_digit(char c)
+{
+	return c >= '0' && c <= '7';
+}
+
 static char *parse_next_string_field(sc_mountinfo_entry * entry,
-				     const char *line, int *offset)
+				     const char *line, size_t * offset)
 {
-	int offset_delta = 0;
-	char *field = &entry->line_buf[0] + *offset;
-	if (line[*offset] == ' ') {
-		// Special case for empty fields which cannot be parsed with %s.
-		*field = '\0';
-		*offset += 1;
-	} else {
-		int nscanned =
-		    sscanf(line + *offset, "%s%n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		*offset += offset_delta;
-		if (line[*offset] == ' ') {
-			*offset += 1;
+	const char *input = &line[*offset];
+	char *output = &entry->line_buf[*offset];
+	const char *s;		// Used for escape logic.
+	size_t input_idx = 0;
+	size_t output_idx = 0;
+
+	// Scan characters until we run out of memory to scan or we find a space.
+	// The kernel uses simple octal escape sequences for the following:
+	// space, tab, newline, backwards slash. Everything else is copied
+	// verbatim.
+	bool stop = false;
+	for (int c; c = input[input_idx], c != '\0' && !stop;
```

Comment:
This for loop is fine but a not as clear as it could be with the assignment and two different checks happening in the conditional. Not a blocker.

Reply:
The assignment is there to abbreviate commonly used memory location. The second stop condition could be refactored out of the equation if we stop using the switch statement and instead use if-then-else tree. Then we could only have one termination condition (c != '\0') and use `break` for additional special-case conditions inline.

Reply:
I reworked the loop.  Please look again.